### ISSUE TITLE
[Clang] `-Wparentheses-equality` produces a false positive for `==` fold expressions in conditions

### ISSUE THREAD

**Issue by Sirraide** [llvm/llvm-project#101863]
The following code (https://godbolt.org/z/rPqP4r55b):
```c++
void foo(auto... args) {
    if (((args == 0) or ...)) {}
}

void f() { 
    foo(3);
}
```
results in a `-Wparentheses-equality` warning:
```
<source>:2:16: warning: equality comparison with extraneous parentheses [-Wparentheses-equality]
    2 |     if (((args == 0) or ...)) {}
      |           ~~~~~^~~~
<source>:6:5: note: in instantiation of function template specialization 'foo<int>' requested here
    6 |     foo(3);
      |     ^
<source>:2:16: note: remove extraneous parentheses around the comparison to silence this warning
    2 |     if (((args == 0) or ...)) {}
      |          ~     ^   ~
<source>:2:16: note: use '=' to turn this equality comparison into an assignment
    2 |     if (((args == 0) or ...)) {}
      |                ^~
      |                =
```
However, the suggestion to remove parentheses here is obviously wrong: neither pair of parentheses can be removed, as that would result in an invalid fold expression (because `args == 0` is not a cast-expression). It seems that, generally speaking, we should not issue this warning if the equality comparison is the operand of a fold expression.

**Comment by Sirraide** [comment]
The culprit seems to be the fact that the AST we get after instantiating this is:
```
`-CompoundStmt 0x289d21d8 <col:24, line:3:1>
  `-IfStmt 0x289d21b8 <line:2:5, col:32>
    |-ParenExpr 0x289d2198 <col:10, col:20> 'bool'
    | `-BinaryOperator 0x289d2178 <col:11, col:19> 'bool' '=='
    |   |-ImplicitCastExpr 0x289d2160 <col:11> 'int' <LValueToRValue>
    |   | `-DeclRefExpr 0x289d2128 <col:11> 'int' lvalue ParmVar 0x289d1e18 'args' 'int'
    |   `-IntegerLiteral 0x289d1a98 <col:19> 'int' 0
    `-CompoundStmt 0x289d1b38 <col:31, col:32>
```
which looks exactly as though the user had written `if ((args == 0)) {}`.

**Comment by a-tarasyuk** [comment]
@Sirraide I looked into it briefly, and it seems that `CondExpr` is transformed in a way that skips `CXXFoldExpr`. Since `TransformCondition` uses `ActOnCondition`, which attempts to validate `(args == 0)` sub exps, it triggers a warning.

https://github.com/llvm/llvm-project/blob/2ab9233f4f393c240c37ef092de09d907fe5c890/clang/lib/Sema/TreeTransform.h#L4411-L4419

https://github.com/llvm/llvm-project/blob/2ab9233f4f393c240c37ef092de09d907fe5c890/clang/lib/Sema/TreeTransform.h#L15429-L15431

https://github.com/llvm/llvm-project/blob/2ab9233f4f393c240c37ef092de09d907fe5c890/clang/lib/Sema/TreeTransform.h#L15521

---

https://github.com/llvm/llvm-project/blob/9ec229dd3ede6368ee5fecf9b9731d8d21d8b1d2/clang/lib/Sema/SemaExpr.cpp#L20284

https://github.com/llvm/llvm-project/blob/9ec229dd3ede6368ee5fecf9b9731d8d21d8b1d2/clang/lib/Sema/SemaExpr.cpp#L20246

https://github.com/llvm/llvm-project/blob/9ec229dd3ede6368ee5fecf9b9731d8d21d8b1d2/clang/lib/Sema/SemaExpr.cpp#L20215-L20224


I'm unsure if `TransformTree` is the appropriate spot for this fix. What approach do you think would be most relevant here? Thanks

**Comment by Sirraide** [comment]
I feel like we need to check whether we’re instantiating a template and skip this check in that case, yeah.

**Comment by Sirraide** [comment]
Maybe checking `inTemplateInstantiation()` would work? If not, we can pass a `bool` from the call to `ActOnCondition` in `TreeTransform` down to this function, but this feels like a bit of a hack...

@erichkeane probably knows what the right way of doing this is.

**Comment by erichkeane** [comment]
Hmm... I think that is too big of a hammer, a lot of these sorts of diagnostics are valid in template cases, the difference here is that it is a necessary paren-expr.

I wonder if when we are doing a fold expression that results in only a single element, we should strip off the extra paren-expr.  Or at least mark it in some way as 'necessary'.

**Comment by erichkeane** [comment]
Addition: `TransformExpr` should eventually go through some sort of `TransformCXXFoldExpr`.  But I find myself wondering if could do a `IgnoreParens` on the component if `getNumExprs` == 1.

**Comment by Sirraide** [comment]
> I wonder if when we are doing a fold expression that results in only a single element, we should strip off the extra paren-expr. Or at least mark it in some way as 'necessary'.

I was gonna say that the first option seems like it would impact source fidelity, but at the same time, unless I’m missing something here, that already went out the window the moment we expanded this to a single `BinaryOperator` ;Þ

> But I find myself wondering if could do a `IgnoreParens` on the component if `getNumExprs` == 1

Thanks, I didn’t think of that, I feel like this might be the right approach (the alternative I guess is to add an `ExpandedFromFoldExpr` bit or sth like that to... somewhere? But that feels like a bit of a hack just for this...)

**Comment by erichkeane** [comment]
> > I wonder if when we are doing a fold expression that results in only a single element, we should strip off the extra paren-expr. Or at least mark it in some way as 'necessary'.
> 
> I was gonna say that the first option seems like it would impact source fidelity, but at the same time, unless I’m missing something here, that already went out the window the moment we expanded this to a single `BinaryOperator` ;Þ
> 
> > But I find myself wondering if could do a `IgnoreParens` on the component if `getNumExprs` == 1
> 
> Thanks, I didn’t think of that, I feel like this might be the right approach (the alternative I guess is to add an `ExpandedFromFoldExpr` bit or sth like that to... somewhere? But that feels like a bit of a hack just for this...)

Right, I think my thoughts are perfectly aligned with yours here.  Yes, loss of source fidelity, but expansions are weird that way anyway.  An `ExpandedFromFoldExpr` bit seems... wasteful if we could get around it by not adding that to the AST.

**Comment by Sirraide** [comment]
> An `ExpandedFromFoldExpr` bit seems... wasteful if we could get around it by not adding that to the AST

Yeah, agreed.

**Comment by a-tarasyuk** [comment]
@erichkeane @Sirraide Thank you for your guidance! Are these the changes https://github.com/llvm/llvm-project/pull/110761 you were referring to?
